### PR TITLE
feat(availability): add beast daemon drain mode

### LINKED ISSUE #1826: feat(availability): add graceful shutdown drain mode
## Description
Implement an LLM-friendly, independently shippable backlog item for **add graceful shutdown drain mode**. This issue should stay narrow enough for one agent/PR while producing durable behavior, tests, and documentation where appropriate.

## Affected locations
- Schedulers, queues, health checks, and capacity policy
- External dependency handling and degradation paths
- Dashboard/status reporting

## Impact
Without this work, future agents and operators will keep rediscovering the same gap, creating avoidable coordination cost, reliability risk, or ambiguous handoffs.

## Recommendation
Implement the scoped behavior with deterministic tests, concise docs, and structured output that PM/liveness tooling can consume.

## Goals
Deliver add graceful shutdown drain mode as a discrete capability or hardening change.
Make the behavior deterministic enough for autonomous workers to verify.
Keep the implementation small enough for a one-issue worker and one PR.
Capture operator/user-facing behavior in docs or help text when applicable.

## Expected outcomes
add graceful shutdown drain mode is available, documented, and test-backed.
Future PM/worker handoffs can point at this issue without needing extra interpretation.
Failure modes are explicit and produce actionable messages instead of silent drift.
The change does not broaden into adjacent backlog work or unrelated refactors.

## Suggested implementation slices
Inventory the current code/docs paths that own this behavior.
Add or update the narrow model/config/schema needed for the feature or guardrail.
Implement the smallest runtime/docs change that satisfies the acceptance criteria.
Add focused fixtures/tests for success, failure, and edge cases.
Update docs/help output and handoff notes so later agents can use it.

## Acceptance criteria
- [ ] There is a dedicated implementation or documentation change for add graceful shutdown drain mode.
- [ ] At least one automated test, fixture, or deterministic verifier covers the intended behavior.
- [ ] At least one negative/edge case is covered so the failure mode is explicit.
- [ ] Docs/help text or inline operator guidance explains how to use or interpret the behavior.
- [ ] The PR description links this issue and summarizes verification evidence.
- [ ] No unrelated production behavior is changed outside the stated scope.

## LLM handoff notes
- Treat this as a one-issue/one-PR unit.
- Start by searching for existing implementations and tests before adding new surfaces.
- Prefer structured outputs and fixtures over prose-only behavior.
- If this overlaps an existing open issue, comment with evidence and stop for PM deduplication instead of creating an unscoped implementation.
- Do not perform broad refactors or change unrelated worker/PM behavior while implementing this ticket.

## Source
Created by the 2026-07-11 fourth strategic backlog expansion run requesting 100 issues. Topic: **Availability**.

### PR DESCRIPTION
**Pull request by djm204**
## Summary
- add beast daemon drain state during graceful shutdown
- make /health return 503 with structured draining metadata while live child runs are being stopped
- reject new mutating beast-control requests with BEAST_DAEMON_DRAINING while keeping authenticated read endpoints available
- document drain-mode behavior in architecture notes

Closes #1826

## Verification
- npm run test --workspace @franken/orchestrator -- tests/integration/beasts/beast-daemon.test.ts
- npm run lint --workspace @franken/orchestrator
- npm run build --workspace @franken/types --workspace @franken/observer --workspace @franken/brain --workspace @franken/critique --workspace @franken/governor --workspace @franken/planner
- npm run typecheck --workspace @franken/orchestrator
- npm run build --workspace @franken/orchestrator

Note: initial orchestrator typecheck before building workspace dependencies failed on missing internal package declarations; building the dependent workspaces resolved it.

### CODE REVIEW

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/http/beast-daemon-app.ts [hunk 1]
```diff
@@ -30,22 +37,42 @@ export function createBeastDaemonApp(options: BeastDaemonAppOptions): Hono {
...
       runs: services.runs.listRuns().length,
-    });
+      draining,
+      ...(draining ? {
+        drain: {
+          enteredAt: options.drainState?.enteredAt,
+          reason: options.drainState?.reason ?? 'shutdown',
+        },
+      } : {}),
+    }, draining ? 503 : 200);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-app.ts:58]: Fixed in 3b2879c9: the chat-server daemon detector now parses the /health body even for HTTP 503 and treats a matching beasts-daemon response with status="draining", ok=false, expected root, and expected pid as a live daemon. That prevents fallback to in-process Beast services while the pidfile owner is still draining.

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 2]
```diff
@@ -120,6 +138,7 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
         return;
       }
       closed = true;
+      drainState.enter('shutdown');
       const shutdownFailures = await stopLiveRuns(services);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts:214]: Fixed in 3b2879c9: mutating beast-control requests are now tracked as active mutations after the drain check. daemon.close() enters drain mode, rejects new mutations, waits for already-active mutating handlers to finish, then snapshots/stops live runs. Added an integration test that holds an in-flight create-run request and verifies shutdown does not stop runs until that mutation completes.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/cli/run.ts [hunk 3]
```diff
@@ -722,18 +722,19 @@ async function isHealthyBeastsDaemonEndpoint(baseUrl: string, expected: { root:
...
-    return record.ok === true
-      && record.service === 'beasts-daemon'
+    const identifiesExpectedDaemon = record.service === 'beasts-daemon'
       && record.root === expected.root
       && record.pid === expected.pid;
+    if (!identifiesExpectedDaemon) {
+      return false;
+    }
+    return (response.ok && record.ok === true)
+      || (response.status === 503 && record.status === 'draining' && record.ok === false);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/cli/run.ts]: Fixed in 7dfce8c8: detected 503/draining daemon health is no longer treated as attachable. Auto-detection now distinguishes attachable/draining/unavailable; for draining pidfile owners chat-server suppresses standalone local Beast services but also does not proxy to the draining daemon, so it avoids both split-brain local work and a dead proxy commitment.

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 4]
```diff
@@ -120,6 +166,8 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
         return;
       }
       closed = true;
+      drainState.enter('shutdown');
+      await drainState.waitForMutations();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts]: Fixed in 7dfce8c8: the in-flight mutation drain wait is now bounded. close() still gives active mutating handlers a chance to finish before the shutdown sweep, but the wait times out so child-run cleanup and process shutdown can continue instead of hanging indefinitely.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/cli/run.ts [hunk 5]
```diff
@@ -1154,8 +1172,9 @@ async function runChatCommandIfRequested(
     const detectedBeastDaemonUrl = !explicitBeastDaemonUrl && beastOperatorToken
       ? await resolveDetectedBeastsDaemonUrl(root, config, logger)
       : undefined;
-    const beastDaemonUrl = explicitBeastDaemonUrl ?? detectedBeastDaemonUrl;
-    const localBeastServices = beastOperatorToken && !beastDaemonUrl
+    const beastDaemonUrl = explicitBeastDaemonUrl ?? (detectedBeastDaemonUrl === 'draining' ? undefined : detectedBeastDaemonUrl);
+    const suppressLocalBeastServices = detectedBeastDaemonUrl === 'draining';
+    const localBeastServices = beastOperatorToken && !beastDaemonUrl && !suppressLocalBeastServices
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/cli/run.ts]: Fixed in fafcafc4: detected draining daemon health now fails chat-server startup instead of starting without Beast routes. That makes the operator retry after the daemon exits and avoids both split-brain local Beast control and a lifetime with /v1/beasts unavailable.

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 6]
```diff
@@ -120,6 +180,8 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
...
       const shutdownFailures = await stopLiveRuns(services);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts]: Fixed in fafcafc4: the in-flight mutation timeout is now treated as a shutdown failure. If mutating handlers do not drain in time, close() closes the HTTP server and throws BeastDaemonDrainTimeoutError instead of racing stopLiveRuns/services.dispose() against active handlers.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 7]
```diff
@@ -120,6 +187,15 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
...
       }
       closed = true;
+      drainState.enter('shutdown');
+      if (!await drainState.waitForMutations()) {
+        if (listening) {
+          const closedServer = closeHttpServer(server);
+          server.closeAllConnections();
+          await closedServer;
+        }
+        throw new BeastDaemonDrainTimeoutError(MUTATION_DRAIN_WAIT_TIMEOUT_MS);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts]: Fixed in 08335ddb: a drain-wait timeout now still performs shutdown cleanup. It closes active HTTP connections, stops/kills live Beast runs, disposes services, releases the pid file, then throws BeastDaemonDrainTimeoutError so callers know shutdown timed out without leaving unmanaged children or a stale daemon pid.

File: packages/franken-orchestrator/src/http/routes/agent-routes.ts [hunk 8]
```diff
@@ -64,6 +65,24 @@ export function agentRoutes(deps: AgentRoutesDeps): Hono {
 
   app.use('/v1/beasts/agents', auth);
   app.use('/v1/beasts/agents/*', auth);
+  app.use('/v1/beasts/agents', async (c, next) => {
+    if (['POST', 'PUT', 'PATCH', 'DELETE'].includes(c.req.method)) {
+      if (deps.drainMutatingRequest) {
+        await deps.drainMutatingRequest(next);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/routes/agent-routes.ts]: Fixed in 08335ddb: removed the agent-route drain re-check. The daemon-level /v1/beasts/* mutation gate remains the single admission point, so admitted agent mutations are counted once and will not be rejected later if shutdown starts while body/rate-limit middleware is still processing.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 9]
```diff
@@ -120,13 +187,26 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
...
       services.dispose();
-      if (listening) {
+      if (listening && !httpServerClosed) {
         const closedServer = closeHttpServer(server);
         server.closeAllConnections();
         await closedServer;
       }
+      if (!drainedMutations) {
+        await releasePidFile(pidFile);
+        throw new BeastDaemonDrainTimeoutError(MUTATION_DRAIN_WAIT_TIMEOUT_MS);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts]: Fixed in 9716de73: drain-timeout shutdown no longer releases the pid file. It still closes HTTP and performs best-effort child-run cleanup/dispose, but keeps the daemon pid claimed when in-flight mutating handlers did not quiesce, preventing a replacement daemon from starting while the old mutation may still be active.

**Comment by djm204** [comment]
Status update: CI is green at head c31ec2f4 and all current Codex review threads are replied/resolved. The final Codex round produced one actionable finding, fixed in 9716de73, followed by shared lesson docs in c31ec2f4. I did not trigger another @codex review because the configured 5-invocation cap for this PR has been reached; a human needs to approve an over-cap final Codex trigger or accept manual review before merge.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 10]
```diff
@@ -120,13 +187,25 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
...
       }
       closed = true;
+      drainState.enter('shutdown');
+      const drainedMutations = await drainState.waitForMutations();
+      let httpServerClosed = false;
+      if (!drainedMutations && listening) {
+        const closedServer = closeHttpServer(server);
+        server.closeAllConnections();
+        await closedServer;
+        httpServerClosed = true;
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts:210]: Fixed in 1de7f4f4e0a633729e5342d618341ca40475558c. On mutation drain timeout, daemon close now closes intake connections and throws BeastDaemonDrainTimeoutError before stopLiveRuns()/dispose(), leaves pid ownership claimed, and resets close state so a later close can finish after the admitted mutation settles. Added regression coverage proving slow mutating requests do not trigger live-run stop/dispose on timeout and that a retry later stops the created run cleanly.

**Comment by djm204** [comment]
@codex review

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/http/beast-daemon-server.ts [hunk 11]
```diff
@@ -120,9 +189,21 @@ export async function startBeastDaemon(options: StartBeastDaemonOptions): Promis
...
+      const drainedMutations = await drainState.waitForMutations(mutationDrainTimeoutMs);
+      let httpServerClosed = false;
+      if (!drainedMutations && listening) {
+        const closedServer = closeHttpServer(server);
+        server.closeAllConnections();
+        await closedServer;
+        listening = false;
+        httpServerClosed = true;
+        closed = false;
+        throw new BeastDaemonDrainTimeoutError(mutationDrainTimeoutMs);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/http/beast-daemon-server.ts]: Fixed in commit 23b4a670. `daemon.close()` now records the drain timeout but does not throw until cleanup finishes: it closes HTTP intake, waits for the admitted mutating route to complete, stops any live child runs created by that route, disposes services, releases the PID file, and then reports `BeastDaemonDrainTimeoutError`. The regression test now covers the in-flight run creation path and verifies `stopLiveRuns` runs before the timeout is surfaced.

**Comment by djm204** [comment]
@codex review
